core/node/rpc/sync/operation.go

```diff
@@ -0,0 +1,269 @@
+package sync
+
+import (
+	"connectrpc.com/connect"
+	"context"
+	"github.com/ethereum/go-ethereum/common"
+	. "github.com/river-build/river/core/node/base"
+	"github.com/river-build/river/core/node/events"
+	"github.com/river-build/river/core/node/nodes"
+	. "github.com/river-build/river/core/node/protocol"
+	"github.com/river-build/river/core/node/rpc/sync/client"
+	"github.com/river-build/river/core/node/shared"
+)
+
+type (
+	// StreamSyncOperation represents a stream sync operation that is currently in progress.
+	StreamSyncOperation struct {
+		// SyncID is the identifier as used with the external client to identify the streams sync operation.
+		SyncID string
+		// ctx is the root context for this subscription, when expires the subscription and all background syncers are
+		// cancelled
+		ctx context.Context
+		// cancel sync operation
+		cancel context.CancelFunc
+		// commands holds incoming requests from the client to add/remove/cancel commands
+		commands chan *subCommand
+		// thisNodeAddress keeps the address of this stream  thisNodeAddress instance
+		thisNodeAddress common.Address
+		// streamCache gives access to streams managed by this thisNodeAddress
+		streamCache events.StreamCache
+		// nodeRegistry is used to get the remote remoteNode endpoint from a thisNodeAddress address
+		nodeRegistry nodes.NodeRegistry
+	}
+
+	// subCommand represents a request to add or remove a stream and ping sync operation
+	subCommand struct {
+		Ctx             context.Context
+		RmStreamReq     *connect.Request[RemoveStreamFromSyncRequest]
+		AddStreamReq    *connect.Request[AddStreamToSyncRequest]
+		PingReq         *connect.Request[PingSyncRequest]
+		DebugDropStream shared.StreamId
+		ReplyErr        chan error
+	}
+)
+
+// NewStreamsSyncOperation initialises a new sync stream operation. It groups the given syncCookies per stream node
+// by its address and subscribes on the internal stream streamCache for local streams.
+//
+// Use the Run method to start syncing.
+func NewStreamsSyncOperation(
+	ctx context.Context,
+	node common.Address,
+	streamCache events.StreamCache,
+	nodeRegistry nodes.NodeRegistry,
+) (*StreamSyncOperation, error) {
+	// make the sync operation cancellable for CancelSync
+	ctx, cancel := context.WithCancel(ctx)
+
+	return &StreamSyncOperation{
+		ctx:             ctx,
+		cancel:          cancel,
+		SyncID:          GenNanoid(),
+		thisNodeAddress: node,
+		commands:        make(chan *subCommand, 10),
+		streamCache:     streamCache,
+		nodeRegistry:    nodeRegistry,
+	}, nil
+}
+
+// Run the stream sync until either sub.Cancel is called or until sub.ctx expired
+func (syncOp *StreamSyncOperation) Run(
+	req *connect.Request[SyncStreamsRequest],
+	res *connect.ServerStream[SyncStreamsResponse],
+) error {
```

Comment:
would it make sense to add this here?
```suggestion
) error {
defer syncOp.cancel()
```